### PR TITLE
Add support for hrepl.

### PR DESCRIPTION
**Pull request by judah**
`hrepl` is a standalone binary that runs REPLS for Bazel Haskell targets.
Its repository is located at: https://github.com/google/hrepl

`hrel` takes a list of labels on the command-line, in contrast to `haskell_repl()`
which requires editing BUILD files maually.  For example:

```
$ hrepl tests/binary-with-lib:lib
```

This change adds specific output groups to `rules_haskell` which allow `hrepl`
to understand its rule types. Internally, `hrepl` runs the following steps:

- Run `bazel build --output_groups=... haskell:toolchain_info` and read the resulting
  protobuf file to get the path to GHC as well as any relevant command-line options.
- Use `bazel query` to figure out exactly what targets to build.
- "bazel build --output_groups=..." to compile the immediate dependencies of
  the targets.
- Read the "LibraryInfo" protobuf file which contains information about
  them (e.g., the paths of the package DBs).
- "bazel build --output_groups=..." the targets to be interpreted.  This makes
  sure their `srcs`, runfiles, C++ dependencies, etc are available, but doesn't
  actually compile them.  Instead, it writes the necessary information to a "CompileInfo"
  protobuf file.
- Run `ghc --interactive`, combining the flags obtained from all the previous steps.

So this change adds two kinds of output groups:

- Library info: For targets that `hrepl` should recognize as dependencies, such as
  `haskell_library`, `haskell_toolchain_library` and `haskell_cabal_library`.
- Compilation info: For targets that `hrepl` should load directly, such as `haskell_library`
  and `haskell_binary`.

This change also makes the `protoc` binary a dependency of all Haskell build rules,
rather than just `haskell_proto_library`.  It writes files in the human-readable
text format, and then uses the `protoc` binary to convert them to the binary proto format.
The text format itself doesn't have a good forwards/backwards compatibility story,
which is important since `hrepl` is a separate, standalone executable.  (For example,
a text format parser will fail if it encounters a field name that it doesn't understand,
which makes it more difficult to deprecate old fields.)

### CODE REVIEW

File: haskell/private/actions/info.bzl [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+            ),
+    )
+    return proto_pb
+
+# Copied from haskell/repl.bzl; not sure if there's a more general solution.
+_COMMON_REPL_OPTIONS = [
+    "-hide-all-packages",
+    "-package=base",
+    "-package=directory",
+]
```
Comment (aherrmann) [inline haskell/private/actions/info.bzl]: Are those needed here? In case of `haskell_repl` these are set because the GHCi script we use to load modules requires those packages. Does `hrepl` rely on these packages, too? If so, maybe it would be better to explicitly add those flags in `hrepl`, to keep the flags close to the code that relies on them.

**Comment by Profpatsch** [comment]
> This change also makes the `protoc` binary a dependency of all Haskell build rules,
> rather than just `haskell_proto_library`

This sounds weird to me, can’t we make that optional?

**Comment by aherrmann** [comment]
> > This change also makes the `protoc` binary a dependency of all Haskell build rules,
> > rather than just `haskell_proto_library`
> 
> This sounds weird to me, can’t we make that optional?

One way that comes to mind is to factor out the rule info output groups into an aspect. That aspect would then depend on protoc, and the Haskell rules wouldn't have to. @judah What to you think?

---

Though, my understanding is that `protoc` will only have to be built if the rule info output groups are actually requested. To test this, I ran the following two commands:
```
$ bazel clean --expunge && bazel build --execution_log_json_file=build_execlog.json //tests/binary-simple
$ bazel clean --expunge && bazel build --execution_log_json_file=info_execlog.json //tests/binary-simple --output_groups=haskell_compile_info,haskell_library_info
```
Looking at the logs, `build_execlog.json` contains no references to `protoc`. So, as I understand, this confirms that `protoc` is only built if the output groups are requested. Is that correct @judah?

**Comment by Profpatsch** [comment]
If it’s only lazily requested, then it’s fine.

File: rule_info/rule_info.proto [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  // nor flags related to dependent packages ("-package-id" and "-package-db").
+  repeated string options = 9;
+
+  // Shared libraries for cc_library rules that this target depends on
+  // transitively.
+  repeated string transitive_cc_shared_libs = 11;
+}
+
+// Information about a built library.
+message LibraryInfo {
```
[the remark this answers is not in the document]
Comment (judah) [reply rule_info/rule_info.proto]: It doesn't currently handle them.  We originally used static archives, but switched a year ago to dynamically-linked GHC to work around bugs in the static loader.   It seems pretty reasonable to me that it could support both.

I filed google/hrepl#7 to track this, though as you've noted it would also require changes to rules_haskell itself.

File: rule_info/rule_info.proto [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  // (namely, those with the "cabal_version" attribute).
+  string package_name = 6;
+
+  // Arguments to GHC for compiling this target.  Does not include source files,
+  // nor flags related to dependent packages ("-package-id" and "-package-db").
+  repeated string options = 9;
+
+  // Shared libraries for cc_library rules that this target depends on
+  // transitively.
+  repeated string transitive_cc_shared_libs = 11;
```
[the remark this answers is not in the document]
Comment (judah) [reply rule_info/rule_info.proto]: Filed google/hrepl#7.

File: haskell/private/haskell_impl.bzl [hunk 4]
```diff
@@ -634,6 +648,26 @@ def haskell_library_impl(ctx):
...
+                # just set it to "".  It won't matter in practice because those rules don't
+                # have runfiles and won't be compiled directly anyway.
+                workspace_name = ctx.workspace_name if hasattr(ctx, "workspace_name") else "",
+                hs = hs,
+                name = ctx.label.name,
+                c = c,
+                posix = posix,
+                cc_info = cc_info,
+                runfiles = default_info.default_runfiles.files
+                if hasattr(default_info, "default_runfiles") and default_info.default_runfiles else depset(),
```
[the remark this answers is not in the document]
Comment (judah) [reply haskell/private/haskell_impl.bzl]: Done.

File: haskell/repositories.bzl [hunk 5]
```diff
@@ -56,6 +56,46 @@ def rules_haskell_dependencies():
...
+
+    if "rules_python" not in excludes:
+        http_archive(
+            name = "rules_python",
+            sha256 = "e5470e92a18aa51830db99a4d9c492cc613761d5bdb7131c04bd92b9834380f6",
+            strip_prefix = "rules_python-4b84ad270387a7c439ebdccfd530e2339601ef27",
+            urls = ["https://github.com/bazelbuild/rules_python/archive/4b84ad270387a7c439ebdccfd530e2339601ef27.tar.gz"],
+        )
+
+
```
[the remark this answers is not in the document]
Comment (judah) [reply haskell/repositories.bzl]: Removed.

File: rule_info/rule_info.proto [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  // The package DBs that we need to explicitly pass to GHC in order to use
+  // this target.  Namely, the DBs needed for all dependencies transitively,
+  // including itself.
+  // Does not include the "core" package DB which is part of the GHC build
+  // itself and produces packages like "base", "ghc", etc.
+  repeated string transitive_package_dbs = 2;
+
+  // Shared libraries containing built Haskell code for this library
+  // and its dependencies.
+  repeated string transitive_haskell_shared_libs = 3;
```
[the remark this answers is not in the document]
Comment (judah) [reply rule_info/rule_info.proto]: Removed.  It's was a detail of our internal implementation, but it's not actually used by the hrepl that we open-sourced.  (And in fact `rules_haskell` wasn't setting it.)

File: haskell/private/haskell_impl.bzl [hunk 7]
```diff
@@ -634,6 +648,26 @@ def haskell_library_impl(ctx):
         coverage_info,
         default_info,
         lib_info,
+        OutputGroupInfo(**dicts.add(
+            compile_info_output_groups(
+                # For haskell_proto_aspect, which doesn't have a ctx.workspace_name,
+                # just set it to "".  It won't matter in practice because those rules don't
+                # have runfiles and won't be compiled directly anyway.
+                workspace_name = ctx.workspace_name if hasattr(ctx, "workspace_name") else "",
```
[the remark this answers is not in the document]
Comment (judah) [reply haskell/private/haskell_impl.bzl]: Done.

File: haskell/private/actions/info.bzl [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+            transitive_package_ids = [lib_info.package_id],
+            transitive_package_dbs =
+                [db.dirname for db in hs_info.package_databases.to_list()]))
+    return {
+        "haskell_transitive_deps": depset(
+          transitive = [
+                hs_info.package_databases,
+                hs_info.interface_dirs,
+                hs_info.dynamic_libraries,
+            ]),
```
[the remark this answers is not in the document]
Comment (judah) [reply haskell/private/actions/info.bzl]: C library dependencies are triggered through `compile_info_output_groups()`; specifically, the `haskell_cdeps_shared_lib` output group.  It should contain transitive C/C++ libraries for all the targets that are being compiled.

In other words,

File: haskell/private/actions/info.bzl [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+    Writes a file with the text format encoding, and then runs "protoc"
+    to convert it to the wire encoding.
+
+    Using the wire encoding allows us use released versions of tools
+    with different versions of the Haskell rules (within reason).
+
+    Args:
+      hs: The current rule context.
+      output_name: The output filename.  The text-encoded file will be named
+        {output_name}.txt, and the encoded file will be named {file_name}.pb.
```
[the remark this answers is not in the document]
Comment (judah) [reply haskell/private/actions/info.bzl]: Done.

File: haskell/private/actions/info.bzl [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+    "is_hs_library",
+)
+
+def write_proto_file(hs, output_name, proto_type, content):
+    """Write an encoded .proto file.
+
+    Writes a file with the text format encoding, and then runs "protoc"
+    to convert it to the wire encoding.
+
+    Using the wire encoding allows us use released versions of tools
```
[the remark this answers is not in the document]
Comment (judah) [reply haskell/private/actions/info.bzl]: Done.

**Comment by judah** [comment]
Yeah, `protoc` is built lazily as @aherrmann observed.  I worded my original comment poorly; it's more that the `com_google_protobuf` repo (which defines how to build `protoc`) now needs to be included in `haskell/repositories.bzl`, whereas before it was only needed if you used `haskell_proto_library`.  But that seems pretty low-overhead, especially since most of the necessary dependencies were already included (like `rules_python`, as pointed out in the review).

My preference would be to keep using output groups rather than aspects. The former is a little more straightforward, while the latter would also require us to make new providers to pass the relevant information between the regular rules and the aspect(s).  (It could also be a bit tricky to integrate with `haskell_proto_aspect`, which itself needs to emit some information and currently does so via the same output groups as the other rules.)

File: haskell/private/actions/info.bzl (hunk 1 above)
Comment (judah) [reply haskell/private/actions/info.bzl]: Yeah, that makes sense to me.  Done.

Comment (judah) [reply haskell/private/actions/info.bzl]: To clarify, it's not that `hrepl` needs them, but `ghci` itself needs `base`:

```
$ stack exec ghci -- -hide-all-packages
<interactive>:1:6: error:
    Not in scope: ‘System.IO.hSetBuffering’
    No module named ‘System.IO’ is imported.
...
```

However I only added `base`, not `directory` which turned out not to be needed; I guess it's only used in your GHCi script.

File: rule_info/rule_info.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (mboes) [reply rule_info/rule_info.proto]: @judah do you remember what bugs those were? We're likewise interested in making the static linker work better for everyone. It would be quite a win: in the common case we could avoid building dynamic libraries altogether.

[the remark this answers is not in the document]
Comment (aherrmann) [reply rule_info/rule_info.proto]: > though as you've noted it would also require changes to rules_haskell itself.

@judah which required changes are you referring to?

`rules_haskell` supports static rts GHC and [sets `-fPIC` and `-fexternal-dynamic-refs` as required](https://github.com/tweag/rules_haskell/blob/d76a16eb2a21b184bfda8dbbc05ee4c1c2239015/haskell/private/actions/compile.bzl#L270) and uses `pic_static_library` of dependencies where available. On Windows that's the default, as the Windows bindist uses static rts mode. On Unix it's optional, currently exposed for nixpkgs provisioned GHC through the [`is_static` attribute](https://api.haskell.build/haskell/nixpkgs.html#haskell_register_ghc_nixpkgs), since there are no static rts GHC bindists available for Unix.

For example, Digital Asset's open source [daml repository](https://github.com/digital-asset/daml/blob/5231bab42399b515c3ae64d8836e2dde3c6a33c7/WORKSPACE#L323) uses static linking in this way. As a test, I've run `hrepl` with this PR's patches on that repository and was able to load a large target with static C library dependencies into GHCi using
```
LIBRARY_PATH=$PWD/bazel-out/k8-opt/bin/compiler/damlc/_ghci_libs-damlc hrepl damlc
```

Looking at `cc_libs` in `compile_info_output_groups`, it doesn't seem to distinguish static and dynamic libraries, so `transitive_cc_shared_libs` will end up holding both. So, maybe just the name is inaccurate.

File: haskell/private/actions/info.bzl (hunk 8 above)
[the remark this answers is not in the document]
Comment (aherrmann) [reply haskell/private/actions/info.bzl]: > In other words,

Did the comment get cropped?

File: haskell/private/actions/info.bzl (hunk 1 above)
Comment (aherrmann) [reply haskell/private/actions/info.bzl]: Thanks!

**Comment by aherrmann** [comment]
> My preference would be to keep using output groups rather than aspects. The former is a little more straightforward, while the latter would also require us to make new providers to pass the relevant information between the regular rules and the aspect(s). 

That's a good point. Let's stick to the output groups then. I'm also fine with a lazy dependency on protoc.

(@judah Sorry, I somehow managed to edit your comment instead of replying. I tried to fix it again.)

**Comment by aherrmann** [comment]
The CI failures should be fixed by rebase on `master` (see https://github.com/tweag/rules_haskell/pull/1214)

File: haskell/private/actions/info.bzl (hunk 8 above)
[the remark this answers is not in the document]
Comment (judah) [reply haskell/private/actions/info.bzl]: Finishing the comment:

In other words, `CompileInfo` provides all files/information about interpreting/running the targets passed on the command-line, and `LibraryInfo` provides files/information about compiling their dependencies.  Since C++ libraries are used at runtime, they're attached to the former.

File: rule_info/rule_info.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (judah) [reply rule_info/rule_info.proto]: Ah, I didn't realize that static libraries got passed through the same field of cc_libs.  I've renamed the field to `transitive_cc_libs`.  (I don't have access to a static GHC at the moment, so we may need to leave that to future work.)

We got hit by a few bugs around the static linker in the transition from 8.0->8.4 and 8.4->8.6, including multiple instances of segfaults on ANN pragmas.  One example was:
https://gitlab.haskell.org/ghc/ghc/issues/14675
But there were a couple others we never tracked down well enough to file a useful bug report (in particular we weren't able to rule out effects of our internal toolchain).

At the time (early 2019), the existing comments on the GHC issues didn't give us a lot of confidence in the static RTS. But it's good to hear you've taken it up in earnest, and your use of `-fexternal-dynamic-refs` sounds promising; we never discovered that flag, and the problem mentioned in that link (`R_X86_64_PC32`) does sound familiar. I'll make a note to try using that in our codebase.

[the remark this answers is not in the document]
Comment (mboes) [reply rule_info/rule_info.proto]: FWIW my understanding is that interest in the static linker from GHC HQ has varied over the years. When support from dynamic libraries was first introduced, I suppose the vision was to eventually retire the static linker, because it's duplicating so much OS functionality. However, the story around dynamic libraries turned out less than ideal. And at this point, I believe even Facebook (who employs @simonmar) might be using a static toolchain, or so https://engineering.fb.com/security/fighting-spam-with-haskell/ would have me believe.

At any rate, the static toolchain is what Digital Asset uses too.

[the remark this answers is not in the document]
Comment (simonmar) [reply rule_info/rule_info.proto]: Actually I believe we're not using the static runtime linker any more at Facebook. The static runtime linker requires all linked objects to fit in the low 2GB of the address space, and we eventually ran into that limit and had to switch to building a standard `.so` and using the system dyanmic linker to load it.

[the remark this answers is not in the document]
Comment (mboes) [reply rule_info/rule_info.proto]: @simonmar very interesting. And thanks for chiming in! So out of curiosity, do you now _only_ use standard .so during your build, or do you still use `-dynamic-too` when building? If the former, that would be nice to speed up the build, but it's not an option for our macOS users, where the Mach-O header size limits are intentionally very low (hence, there's only so many dynamic dependencies one can have).

File: haskell/private/actions/info.bzl [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+        runfiles: A depset of Files.
+
+    Returns:
+      A dict whose keys are output groups and values are depsets of Files.
+    """
+    (ghci_extra_libs, ghc_env) = get_ghci_extra_libs(hs, posix, cc_info)
+    cc_libs = [lib for lib in ghci_extra_libs.to_list()
+                if not is_hs_library(get_lib_name(lib)) and get_lib_name(lib) != "ffi"]
+    return {
+        "haskell_cdeps_shared_lib": depset(cc_libs),
```
Comment (aherrmann) [inline haskell/private/actions/info.bzl]:
```suggestion
        "haskell_cdeps_lib": depset(cc_libs),
```
Since `cc_libs` doesn't distinguish static and shared libraries. I.e. same reasoning as [here](https://github.com/tweag/rules_haskell/pull/1210/files#r367520407).

File: haskell/private/actions/info.bzl (hunk 8 above)
[the remark this answers is not in the document]
Comment (aherrmann) [reply haskell/private/actions/info.bzl]: Understood, thanks for explaining! Since `CcInfo` carries transitive C library dependencies this will also contain C libraries introduced by other Haskell dependencies. Makes sense.

File: haskell/private/actions/info.bzl (hunk 11 above)
Comment (judah) [reply haskell/private/actions/info.bzl]: Done. (Renamed to `haskell_cdep_libs` to be a little more like the other providers.)

Comment (aherrmann) [reply haskell/private/actions/info.bzl]: Thanks!

**Comment by judah** [comment]
@aherrmann yes, let's do the testing in a follow-up PR.  The `hrepl` repo itself has tests which are (sort of) driven by `bazel`:
https://github.com/google/hrepl/blob/master/hrepl/tests/run_tests.sh
https://github.com/google/hrepl/blob/d946c54471fa5f1ca7dc5b8baf0db6a19429b319/hrepl/tests/BUILD.bazel#L5

They work by pointing `bazel` to the current directory but with `--output_base` pointing to a temporary directory.  It uses `--test_env` to inject those values into a given `bazel test` invocation.  However, because they need to download and install the GHC bindist, etc. in a new location (due to different `output_base`), they take ~10m to run on my laptop, and probably more in CircleCI.

I can think of a few possiblities:
1. Run hrepl's tests in CI (or maybe a subset of them).
2. Run a simplified version of hrepl's tests in CI that doesn't try to be as hermetic, and reuses the same `output_base` as the rest of the build. It should be much faster as a result.  It would still miss testing some edge case situations, but we'd get that covered by the `hrepl` repo's own testing.
3. Do something much simpler, similar to what `RunTests.hs` does [now](https://github.com/tweag/rules_haskell/blob/d3ca7094c727d9dfda464c178bb0e4ae0171baa5/tests/RunTests.hs#L49).  For example, instead of `bazel run //...@repl -- ...` do `bazel build @hrepl//hrepl; bazel-bin/external/hrepl/hrepl/hrepl //... ...`

**Comment by aherrmann** [comment]
I took a look at CI
* the Azure failure is unrelated to this PR and fixed by https://github.com/tweag/rules_haskell/pull/1220.
* on buildkite linting failed and is fixed by `bazel run //:buildifier-fix`.
* on buildkite the "example" test-case failed, because it collided with the `@zip` external workspace due to the `protoc` dependency. 

I've pushed fixes to this PR.
